### PR TITLE
support custom root types

### PR DESCRIPTION
**Pull request by koxudaxi**
## Change Summary

support custom root types that support a model that does not have any field exclude ``__root__``.

example: 
```python
from typing import List
import json
from pydantic import BaseModel
from pydantic.schema import schema

class Pets(BaseModel):
    __root__: List[str]

print(Pets(__root__=['dog', 'cat']))
# > Pets __root__=['dog', 'cat']

print(Pets(*['dog', 'cat']))
# > Pets __root__=['dog', 'cat']

print(Pets(*['dog', 'cat']).schema())
# > {'title': 'Pets', 'type': 'array', 'items': {'type': 'string'}}

pets_schema = schema([Pets])
print(json.dumps(pets_schema, indent=2))

# {
#  "definitions": {
#    "Pets": {
#      "title": "Pets",
#      "type": "array",
#      ...


```
## Related issue number

https://github.com/samuelcolvin/pydantic/issues/507

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `HISTORY.rst` has been updated
  * this pull request number `#628`
  * include github username `@koxudaxi`

### CODE REVIEW

File: docs/index.rst [hunk 1]
```diff
@@ -466,6 +466,18 @@ Outputs:
 .. literalinclude:: examples/schema3.json
 
 
+Pydantic models can be set custom root types to support a model that does not have any field exclude ``__root__``.
```
Comment (samuelcolvin) [inline docs/index.rst]: Can you make this a separate section of the docs. This is not just about schema generation.

File: pydantic/main.py [hunk 2]
```diff
@@ -186,7 +186,7 @@ def __new__(mcs, name, bases, namespace):
             for ann_name, ann_type in annotations.items():
                 if is_classvar(ann_type):
                     class_vars.add(ann_name)
-                elif not ann_name.startswith('_') and ann_name not in namespace:
+                elif (not ann_name.startswith('_') or '__root__' == ann_name) and ann_name not in namespace:
```
Comment (samuelcolvin) [inline pydantic/main.py]: we do this at least three times, can we have a function for it.

File: pydantic/main.py [hunk 3]
```diff
@@ -233,16 +235,27 @@ class BaseModel(metaclass=MetaModel):
...
         __config__: Type[BaseConfig] = BaseConfig
+        __root__: List[Any] = []
         _json_encoder: Callable[[Any], Any] = lambda x: x
         _schema_cache: 'DictAny' = {}
 
     Config = BaseConfig
     __slots__ = ('__values__', '__fields_set__')
 
-    def __init__(self, **data: Any) -> None:
+    def __init__(self, *__root__: Any, **data: Any) -> None:
```
Comment (samuelcolvin) [inline pydantic/main.py]: Humm, I've very unwilling to change the signature of `__init__`. 

Would it be possible to accept `__root__` directly to `parse_obj` only?

File: docs/index.rst (hunk 1 above)
Comment (koxudaxi) [reply docs/index.rst]: I see. I have moved the content to around types sections.

File: pydantic/main.py (hunk 2 above)
Comment (koxudaxi) [reply pydantic/main.py]: I have created ``is_valid_field function`` for checking a field name.

File: pydantic/main.py (hunk 3 above)
Comment (koxudaxi) [reply pydantic/main.py]: I have reverted the  ``__init__`` method to master.
And I add a ``__root__`` argument in ``parse_obj``.
However, We can pass ``__root__`` object to ``__init__`` as a keyword argument.
```python
Model(__root__='a')
# or
Model.parse_obj(__root__='a')
```

File: docs/index.rst [hunk 4]
```diff
@@ -669,6 +669,20 @@ to get validators to parse and validate the input data.
 
 (This script is complete, it should run "as is")
 
+Custom Root Types
+.................
+
+Pydantic models can be set custom root types to support a model that does not have any field exclude ``__root__``.
```
Comment (samuelcolvin) [inline docs/index.rst]:
```suggestion
Pydantic models which do not represent an "object" in JSON parlance (``dict`` in python) can have a custom
root type defined via the ``__root__`` field. The root type can of any time: list, float, int etc.
```

File: docs/index.rst [hunk 5]
```diff
@@ -669,6 +669,20 @@ to get validators to parse and validate the input data.
...
+
+You can define custom root types as a type hint of  ``__root__``'s field.
```
Comment (samuelcolvin) [inline docs/index.rst]:
```suggestion
The root type can be defined via the type hint on the ``__root__`` field.
```

File: docs/index.rst [hunk 6]
```diff
@@ -669,6 +669,20 @@ to get validators to parse and validate the input data.
...
+Also, You can give custom root values to the ``__init__`` method or ``parse_obj`` as a keyword argument.
```
Comment (samuelcolvin) [inline docs/index.rst]:
```suggestion
The root value can be passed to model ``__init__`` via the ``__root__`` argument as the first and only argument
to ``parse_obj``.
```

File: pydantic/main.py [hunk 7]
```diff
@@ -233,6 +241,7 @@ class BaseModel(metaclass=MetaModel):
         __fields__: Dict[str, Field] = {}
         __validators__: Dict[str, AnyCallable] = {}
         __config__: Type[BaseConfig] = BaseConfig
+        __root__: List[Any] = []
```
Comment (samuelcolvin) [inline pydantic/main.py]: I thought `__root__` doesn't have to be a list?

File: pydantic/main.py [hunk 8]
```diff
@@ -211,6 +217,8 @@ def __new__(mcs, name, bases, namespace):
                         config=config,
                     )
 
+        if '__root__' in fields and len(fields) > 1:
+            raise ValueError('__root__ cannot be mixed with other fields')
```
Comment (samuelcolvin) [inline pydantic/main.py]: add `_custom_root_type: bool` to `new_namespace` below.

File: pydantic/main.py [hunk 9]
```diff
@@ -324,9 +334,17 @@ def json(
         )
 
     @classmethod
-    def parse_obj(cls: Type['Model'], obj: Mapping[Any, Any]) -> 'Model':
+    def parse_obj(
+        cls: Type['Model'], obj: Optional[Mapping[Any, Any]] = None, __root__: Optional[Any] = None
```
Comment (samuelcolvin) [inline pydantic/main.py]: simplify this to

```py
def parse_obj(cls: Type['Model'], obj: Any) -> 'Model'
```

File: pydantic/main.py [hunk 10]
```diff
@@ -324,9 +334,17 @@ def json(
...
+    ) -> 'Model':
+
+        if __root__:
+            return cls(__root__=__root__)
+
         if not isinstance(obj, dict):
             try:
+                if obj is None:
```
Comment (samuelcolvin) [inline pydantic/main.py]: change this to:

```py
if not instanstance(obj, dict):
    if cls._custom_root_type:
        obj = {'__root__': obj}
    else:
        ...
```

File: tests/test_main.py [hunk 11]
```diff
@@ -649,3 +649,38 @@ class Config:
     m = MyModel(extra_key='extra')
     assert m.dict() == {'a': None, 'extra_key': 'extra'}
     assert m.dict(by_alias=True) == {'alias_a': None, 'extra_key': 'extra'}
+
+
+def test_root():
+    class MyModel(BaseModel):
+        __root__: str
+
+    m = MyModel(__root__='a')
```
Comment (samuelcolvin) [inline tests/test_main.py]: please test with `parse_obj` too.

File: pydantic/main.py (hunk 10 above)
Comment (samuelcolvin) [reply pydantic/main.py]: I think we should also raise an error if the type of `__root__` is a dict.

File: pydantic/main.py (hunk 8 above)
Comment (koxudaxi) [reply pydantic/main.py]: I have added it.

File: pydantic/main.py (hunk 7 above)
Comment (koxudaxi) [reply pydantic/main.py]: Sorry, it's my mistake. I have changed ``List[Any]`` to ``Any``.

File: pydantic/main.py (hunk 9 above)
Comment (koxudaxi) [reply pydantic/main.py]: ok, I just changed it.

File: pydantic/main.py (hunk 10 above)
Comment (koxudaxi) [reply pydantic/main.py]: I have changed the condition and added to raise an exception.

File: tests/test_main.py (hunk 11 above)
Comment (koxudaxi) [reply tests/test_main.py]: I had made tests for ``parse_obj`` in ``tests/tset_parse.py``
Would you please check it?
https://github.com/samuelcolvin/pydantic/pull/628/files#diff-7b34c8bd665994bbdcf056c03622d4a8R45

File: pydantic/main.py [hunk 12]
```diff
@@ -330,13 +341,19 @@ def json(
...
-        if not isinstance(obj, dict):
-            try:
-                obj = dict(obj)
-            except (TypeError, ValueError) as e:
-                exc = TypeError(f'{cls.__name__} expected dict not {type(obj).__name__}')
-                raise ValidationError([ErrorWrapper(exc, loc='__obj__')]) from e
+    def parse_obj(cls: Type['Model'], obj: Any) -> 'Model':
+        if isinstance(obj, dict):
+            if cls._custom_root_type:
+                raise TypeError('custom root type cannot allow dict')
```
[the remark this answers is not in the document]
Comment (koxudaxi) [reply pydantic/main.py]: OK, I have moved it to ``__new__`` in MetaModel. 
Also, I change the validation type from ``dict`` to ``Mapping``. If it's wrong, then please tell me correct.

**Comment by koxudaxi** [comment]
> sorry, reviewed this days ago and somehow forgot to press submit.

No problem, Thank you for reviewing my code.

**Comment by mdsrosa** [comment]
Hey @koxudaxi @samuelcolvin. I have a question that is related to this PR. 

Given the following use case:

```python
from typing import List
from datetime import datetime
from pydantic import BaseModel

class LotteryResults(BaseModel):
    __root__: List[int]

class Lottery(BaseModel):
    when: datetime
    numbers: LotteryResults

numbers = LotteryResults.parse_obj([10, 20, 30, 12 , 34]
lottery = Lottery(when=datetime.now(), numbers=numbers)
```

Running that code as is, when I call the `.json()` on `lottery` I get the following result:
```python
In [7]: lottery.json()
Out[7]: '{"when": "2020-05-21T14:09:14.652009", "numbers": {"__root__": [10, 20, 30, 12, 34]}}'
```
Is it possible to have the `LotterResults` return as the list of numbers? That would be much easier to consume without any logic that depends on a `__root__` field existing. I would like to get the following result instead:

```python
In [7]: lottery.json()
Out[7]: '{"when": "2020-05-21T14:09:14.652009", "numbers": [10, 20, 30, 12, 34]}'
```

By the way, Pydantic is awesome!

**Comment by koxudaxi** [comment]
@mdsrosa 

I paste a workaround.
Btw, I think you should create a new issue for this question. :memo: 

```python
import json
from typing import Any
from typing import List
from datetime import datetime
from pydantic import BaseModel
from pydantic.class_validators import ROOT_KEY


def unpack_custom_root(value: Any):
    if isinstance(value, dict):
        if ROOT_KEY in value:
            return value[ROOT_KEY]
        return {k: unpack_custom_root(v) for k, v in value.items()}
    return value


def custom_root_json_dumps(value, *, default):
    return json.dumps(unpack_custom_root(value), default=default)


class LotteryResults(BaseModel):
    __root__: List[int]


class Lottery(BaseModel):
    when: datetime
    numbers: LotteryResults

    class Config:
        json_dumps = custom_root_json_dumps


numbers = LotteryResults.parse_obj([10, 20, 30, 12, 34])
lottery = Lottery(when=datetime.now(), numbers=numbers)

print(lottery.json())
#  {"when": "2020-05-22T00:32:10.811364", "numbers": [10, 20, 30, 12, 34]}

```

**Pull request merged**